src/ray/gcs/store_client/redis_store_client.h

```diff
@@ -130,27 +216,32 @@ class RedisStoreClient : public StoreClient {
   //
   // \return The requests to send.
   std::vector<std::function<void()>> TakeRequestsFromSendingQueue(
-      const std::vector<std::string> &keys) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
-
-  Status DoPut(const std::string &key,
-               const std::string &data,
-               bool overwrite,
-               std::function<void(bool)> callback);
+      const std::vector<RedisConcurrencyKey> &keys) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
 
-  Status DeleteByKeys(const std::vector<std::string> &keys,
+  Status DeleteByKeys(const std::string &table_name,
+                      const std::vector<std::string> &keys,
                       std::function<void(int64_t)> callback);
 
   // Send the redis command to the server. This method will make request to be
-  // serialized for each key in keys. At a given time, only one request for a key
-  // will be in flight.
+  // serialized for each key in keys. At a given time, only one request for a {table_name,
+  // key} will be in flight.
   //
-  // \param keys The keys in the request.
+  // \param keys Used as concurrency key.
   // \param args The redis commands
   // \param redis_callback The callback to call when the reply is received.
-  void SendRedisCmd(std::vector<std::string> keys,
-                    std::vector<std::string> args,
-                    RedisCallback redis_callback);
-
+  void SendRedisCmdWithKeys(std::vector<std::string> keys,
+                            RedisCommand command,
+                            RedisCallback redis_callback);
+
+  // Conveinence method for SendRedisCmdWithKeys with keys = command.args.
+  // Reason for this method: if you call SendRedisCmd(command.args, std::move(command)),
+  // it's UB because C++ don't have arg evaluation order guarantee, hence command.args
+  // may become empty.
+  void SendRedisCmd(RedisCommand command, RedisCallback redis_callback);
```

Comment:
```suggestion
  void SendRedisCmdArgsAsKeys(RedisCommand command, RedisCallback redis_callback);
```

I'm a bit worried that users will misuse these two functions without reading the comment. This rename should make it a bit clear.

Or should we still just have a single version of SendRedisCmd and force users to always specify keys (they need to manually do what SendRedisCmd does in this PR)?

Reply:
updated